File: src/infrastructure/spreadsheet/baseSheet.js
/* exported BaseSheet */

class BaseSheet {
  constructor(sheetName, startRow = 1, startColumn = 1){
    const spreadsheet = SpreadsheetApp.getActiveSpreadsheet();
    this.sheet = spreadsheet.getSheetByName(sheetName);
    if (!this.sheet) {
      throw new Error(`シート "${sheetName}" が見つかりません`);
    }
    this.lastRow = this.sheet.getLastRow();
    this.headerRow = startRow;
    this.startRow = startRow + 1;
    this.startColumn = startColumn;
    
    // 列名->index 解決に使うヘッダー（startRow行のみ）
    const headerRaw = this.sheet.getRange(this.headerRow, startColumn, 1, 100).getValues()[0] || [];
    this._headersPrimaryRaw = headerRaw;
    this._headersPrimary = headerRaw.map(h => String(h ?? '').trim());
    this._headerIndexMap = new Map();
    for (let i = 0; i < this._headersPrimary.length; i++) {
      const key = this._headersPrimary[i];
      if (!key) continue;
      if (!this._headerIndexMap.has(key)) {
        this._headerIndexMap.set(key, i);
      }
    }
    
    // データ範囲を計算（startRowから最後の行まで、100列分）
    const numRows = Math.max(0, this.lastRow - this.startRow + 1);
    const raw = numRows > 0 ? this.sheet.getRange(this.startRow, startColumn, numRows, 100).getValues() : [];

    this.data = raw.map((values, i) => new BaseRow(values, (name) => this._getColumnIndexByName(name), this.startRow + i));
    // getActiveRowData() が this.data を選択行で上書きするため、全行データも保持しておく
    this.allData = this.data;
  }

  _getColumnIndexByName(columnName) {
    const key = String(columnName ?? '').trim();
    const idx = this._headerIndexMap.get(key);
    if (idx !== undefined) {
      return idx;
    }
    throw new Error(`列 "${key}" が見つかりません。ヘッダー: ${JSON.stringify(this._headersPrimary.filter(Boolean))}`);
  }

  getActiveRowData(){
    const selectedRowNumbers = this._getSelectedRowNumbers(this.sheet);
    const sortedRowNumbers = Array.from(selectedRowNumbers).sort((a, b) => a - b);
    const selectedRows = [];
    const finalRowNumbers = [];
    const skippedRows = [];

    for (const rowNum of sortedRowNumbers) {
      const isHidden = this.sheet.isRowHiddenByFilter(rowNum) || this.sheet.isRowHiddenByUser(rowNum);
      console.log(`行番号 ${rowNum}: フィルタ非表示=${this.sheet.isRowHiddenByFilter(rowNum)}, 手動非表示=${this.sheet.isRowHiddenByUser(rowNum)}, 判定=${isHidden ? 'スキップ' : '対象'}`);
      if (isHidden) {
        skippedRows.push({row: rowNum, reason: '非表示'});
        continue;
      }

      // データ配列は0始まり、行番号は1始まり、startRowから始まるため
      const dataIndex = rowNum - this.startRow;
      if (dataIndex < 0 || dataIndex >= this.data.length) {
        skippedRows.push({row: rowNum, reason: `データ範囲外（開始行: ${this.startRow}, データ行数: ${this.data.length})`});
        continue;
      }

      const rowData = this.data[dataIndex];

      selectedRows.push(rowData);
      finalRowNumbers.push(rowNum);
    }

    // データ検証: 選択された行が存在しない場合はエラーを投げる
    if (selectedRows.length === 0) {
      const skippedInfo = skippedRows.length > 0 
        ? `\nスキップされた行: ${skippedRows.map(s => `行${s.row}(${s.reason})`).join(', ')}`
        : '';
      throw new Error(`選択された行がありません。すべての選択行が非表示またはデータ範囲外です。${skippedInfo}`);
    }

    // インスタンスの状態を更新
    this.data = selectedRows;
    this.rowNumbers = finalRowNumbers;

    console.log(`選択された行: [${finalRowNumbers.join(', ')}]`);

    return selectedRows;
  }

  _getSelectedRowNumbers(activeSheet) {
    const selectedRowNumbers = new Set();
    
    // 複数の範囲が選択されている場合
    const activeRangeList = activeSheet.getActiveRangeList();
    if (activeRangeList && activeRangeList.getRanges().length > 0) {
      const ranges = activeRangeList.getRanges();
      for (const range of ranges) {
        this._addSelectedRowNumbers(selectedRowNumbers, range);
      }
    } else {
      // 単一の範囲が選択されている場合
      const activeRange = activeSheet.getActiveRange();
      if (activeRange) {
        this._addSelectedRowNumbers(selectedRowNumbers, activeRange);
      }
    }
    
    console.log(`Selected row numbers: ${Array.from(selectedRowNumbers).join(', ') || '(なし)'}`);
    return selectedRowNumbers;
  }

  _addSelectedRowNumbers(selectedRowNumbers, activeRange){
      const startRow = activeRange.getRow();
      const numRows = activeRange.getNumRows();
      console.log(`Adding rows: rownums${startRow} to ${startRow + numRows - 1}`);
      for (let i = 0; i < numRows; i++) {
        selectedRowNumbers.add(startRow + i);
      }
      return selectedRowNumbers;
  }

  writeCell(rowNum, columnNum, value){
    try {
      if (typeof value === 'object' && value.type === 'formula') {
        this.sheet.getRange(rowNum, columnNum).setFormula(value.value);
      } else {
        this.sheet.getRange(rowNum, columnNum).setValue(value);
      }
      console.log(`${rowNum}行目の${columnNum}に書き込みました`);
    } catch (e) {
      console.warn(`${rowNum}行目の${columnNum}への書き込みに失敗しました: ${e.message}`);
      throw e;
    }
  }

  /**
   * columnName列へ、各行に対して関数で計算した値を書き込む。
   * - valueFunc(row, index) が返す値:
   *   - そのまま setValue する値
   *   - {type:'formula', value:'=...'} なら setFormula
   *   - null/undefined ならスキップ
   */
  writeColumnByFunc(columnName, valueFunc) {
    const columnNum = this._getColumnIndexByName(columnName) + 1;
    let successCount = 0;

    for (let i = 0; i < this.data.length; i++) {
      const row = this.data[i];
      const rowNum = row && row.rowNumber ? row.rowNumber : null;
      if (!rowNum) continue;

      const value = valueFunc(row, i);
      if (value === null || value === undefined) continue;

      this.writeCell(rowNum, columnNum, value);
      successCount++;
    }

    console.log(`${columnName}を${successCount}行に書き込みました(by func)`);
    return successCount;
  }
}

File: python/src/infrastructure/spreadsheet/base_sheet.py
from __future__ import annotations

import logging
from typing import Any, Callable, TYPE_CHECKING

import gspread

from infrastructure.spreadsheet.base_row import BaseRow
from infrastructure.spreadsheet.base_sheets_repository import BaseSheetsRepository

if TYPE_CHECKING:
    from gspread import Worksheet

logger = logging.getLogger(__name__)


class BaseSheet:
    def __init__(
        self,
        repo: BaseSheetsRepository,
        sheet_id: str,
        sheet_name: str,
        header_row: int = 1,
    ) -> None:
        self._repo = repo
        self._worksheet: Worksheet = repo.open_worksheet(sheet_id, sheet_name)
        self.header_row = header_row
        self.start_row = header_row + 1

        all_values = self._worksheet.get_all_values()
        if len(all_values) < header_row:
            raise ValueError(f'シート "{sheet_name}" にヘッダー行({header_row})がありません')

        header_raw = all_values[header_row - 1]
        self._headers: list[str] = [str(h).strip() for h in header_raw]
        self._header_index_map: dict[str, int] = {}
        for i, key in enumerate(self._headers):
            if key and key not in self._header_index_map:
                self._header_index_map[key] = i

        data_rows = all_values[header_row:]
        self.all_data: list[BaseRow] = [
            BaseRow(
                values=row,
                column_index_resolver=self._get_column_index_by_name,
                row_number=self.start_row + i,
            )
            for i, row in enumerate(data_rows)
        ]
        self.data: list[BaseRow] = list(self.all_data)

    def _get_column_index_by_name(self, column_name: str) -> int:
        key = str(column_name).strip()
        idx = self._header_index_map.get(key)
        if idx is not None:
            return idx
        valid = [h for h in self._headers if h]
        raise ValueError(f'列 "{key}" が見つかりません。ヘッダー: {valid}')

    def get_rows_by_numbers(self, row_numbers: list[int]) -> list[BaseRow]:
        row_set = set(row_numbers)
        rows = [r for r in self.all_data if r.row_number in row_set]
        self.data = rows
        return rows

    def filter(self, column_name: str, values: list[Any]) -> list[BaseRow]:
        col_idx = self._get_column_index_by_name(column_name)
        str_values = {str(v) for v in values}
        filtered = [r for r in self.all_data if str(r[col_idx]) in str_values]
        self.data = filtered
        logger.info("%sでフィルタリング: %d行が見つかりました", column_name, len(filtered))
        return filtered

    def write_cell(self, row_num: int, column_num: int, value: Any) -> None:
        self._worksheet.update_cell(row_num, column_num, value)
        logger.info("%d行目の%d列に書き込みました", row_num, column_num)

    def write_formula(self, row_num: int, column_num: int, formula: str) -> None:
        cell_label = gspread.utils.rowcol_to_a1(row_num, column_num)
        self._worksheet.update_acell(cell_label, formula)
        logger.info("%d行目の%d列に数式を書き込みました", row_num, column_num)

    def write_column_by_func(
        self, column_name: str, value_func: Callable[[BaseRow, int], Any | None]
    ) -> int:
        col_num = self._get_column_index_by_name(column_name) + 1
        count = 0
        for i, row in enumerate(self.data):
            value = value_func(row, i)
            if value is None:
                continue
            if isinstance(value, dict) and value.get("type") == "formula":
                self.write_formula(row.row_number, col_num, value["value"])
            else:
                self.write_cell(row.row_number, col_num, value)
            count += 1
        logger.info("%sを%d行に書き込みました", column_name, count)
        return count
